Index: gnome-settings-daemon-3.6.1/plugins/media-keys/gsd-media-keys-manager.c
===================================================================
--- gnome-settings-daemon-3.6.1.orig/plugins/media-keys/gsd-media-keys-manager.c	2012-10-26 10:19:49.447004173 +0200
+++ gnome-settings-daemon-3.6.1/plugins/media-keys/gsd-media-keys-manager.c	2012-10-26 10:20:09.391005138 +0200
@@ -2291,6 +2291,9 @@
         g_debug ("Starting media_keys manager");
         gnome_settings_profile_start (NULL);
 
+
+        gvc_mixer_control_open (manager->priv->volume);
+
         manager->priv->settings = g_settings_new (SETTINGS_BINDING_DIR);
         g_signal_connect (G_OBJECT (manager->priv->settings), "changed",
                           G_CALLBACK (gsettings_changed_cb), manager);
@@ -2395,8 +2398,6 @@
                           G_CALLBACK (on_control_stream_removed),
                           manager);
 
-        gvc_mixer_control_open (manager->priv->volume);
-
         gnome_settings_profile_end ("gvc_mixer_control_new");
 
         manager->priv->start_idle_id = g_idle_add ((GSourceFunc) start_media_keys_idle_cb, manager);
