Index: gnome-settings-daemon-3.6.3/configure.ac
===================================================================
--- gnome-settings-daemon-3.6.3.orig/configure.ac	2012-11-19 19:54:25.827630032 +0100
+++ gnome-settings-daemon-3.6.3/configure.ac	2012-11-19 19:54:35.111630482 +0100
@@ -53,6 +53,7 @@
 PA_REQUIRED_VERSION=0.9.16
 LIBWACOM_REQUIRED_VERSION=0.6
 UPOWER_REQUIRED_VERSION=0.9.11
+APPINDICATOR_REQUIRED_VERSION=0.3.0
 IBUS_REQUIRED_VERSION=1.4.99
 
 EXTRA_COMPILE_WARNINGS(yes)
@@ -149,6 +150,32 @@
 
 PKG_CHECK_MODULES(AUTOMOUNT, x11 kbproto)
 
+dnl ---------------------------------
+dnl - Application indicator
+dnl ---------------------------------
+
+AC_ARG_ENABLE([appindicator],
+	AS_HELP_STRING([--enable-appindicator[=@<:@no/auto/yes@:>@]],[Build support for application indicators]),
+	[enable_appindicator=$enableval],
+	[enable_appindicator="auto"])
+
+
+if test x$enable_appindicator = xauto ; then
+	PKG_CHECK_EXISTS(appindicator3-0.1 >= $APPINDICATOR_REQUIRED_VERSION,
+		[enable_appindicator="yes"],
+		[enable_appindicator="no"])
+fi
+
+if test x$enable_appindicator = xyes ; then
+	PKG_CHECK_MODULES(APPINDICATOR,
+		[appindicator3-0.1 >= $APPINDICATOR_REQUIRED_VERSION],
+		[AC_DEFINE(HAVE_APPINDICATOR, 1, [Have AppIndicator])])
+fi
+
+AM_CONDITIONAL(HAVE_APPINDICATOR, test x$enable_appindicator = xyes)
+AC_SUBST(APPINDICATOR_CFLAGS)
+AC_SUBST(APPINDICATOR_LIBS)
+
 dnl ---------------------------------------------------------------------------
 dnl - background
 dnl ---------------------------------------------------------------------------
@@ -569,6 +596,7 @@
         LCMS DICT support:        ${have_new_lcms}
         IBus support:             ${enable_ibus}
         Libnotify support:        ${have_libnotify}
+	App indicator support:    ${enable_appindicator}
         PackageKit support:       ${have_packagekit}
         Smartcard support:        ${have_smartcard_support}
         Cups support:             ${enable_cups}
Index: gnome-settings-daemon-3.6.3/plugins/a11y-keyboard/gsd-a11y-keyboard-manager.c
===================================================================
--- gnome-settings-daemon-3.6.3.orig/plugins/a11y-keyboard/gsd-a11y-keyboard-manager.c	2012-11-14 13:19:27.000000000 +0100
+++ gnome-settings-daemon-3.6.3/plugins/a11y-keyboard/gsd-a11y-keyboard-manager.c	2012-11-19 19:54:25.991630041 +0100
@@ -41,6 +41,10 @@
 #include <X11/XKBlib.h>
 #include <X11/extensions/XKBstr.h>
 
+#ifdef HAVE_APPINDICATOR
+#include <libappindicator/app-indicator.h>
+#endif
+
 #include "gnome-settings-profile.h"
 #include "gsd-a11y-keyboard-manager.h"
 #include "gsd-a11y-preferences-dialog.h"
@@ -61,7 +65,11 @@
         GtkWidget        *stickykeys_alert;
         GtkWidget        *slowkeys_alert;
         GtkWidget        *preferences_dialog;
+#ifdef HAVE_APPINDICATOR
+        AppIndicator *app_indicator;
+#else
         GtkStatusIcon    *status_icon;
+#endif
 
         GSettings        *settings;
 
@@ -427,11 +435,24 @@
         /* for now, show if accessx is enabled */
         show = g_settings_get_boolean (manager->priv->settings, "enable");
 
+#ifdef HAVE_APPINDICATOR
+        if (!show && manager->priv->app_indicator == NULL)
+                return;
+	
+        gsd_a11y_keyboard_manager_ensure_status_icon (manager);
+	if (show)
+		app_indicator_set_status (manager->priv->app_indicator,
+					  APP_INDICATOR_STATUS_ACTIVE);
+	else
+		app_indicator_set_status (manager->priv->app_indicator,
+					  APP_INDICATOR_STATUS_PASSIVE);
+#else
         if (!show && manager->priv->status_icon == NULL)
                 return;
 
         gsd_a11y_keyboard_manager_ensure_status_icon (manager);
         gtk_status_icon_set_visible (manager->priv->status_icon, show);
+#endif
 }
 
 static void
@@ -509,9 +530,11 @@
         message = _("You just held down the Shift key for 8 seconds.  This is the shortcut "
                     "for the Slow Keys feature, which affects the way your keyboard works.");
 
+#ifndef HAVE_APPINDICATOR
         if (manager->priv->status_icon == NULL || ! gtk_status_icon_is_embedded (manager->priv->status_icon)) {
                 return FALSE;
         }
+#endif
 
         if (manager->priv->slowkeys_alert != NULL) {
                 gtk_widget_destroy (manager->priv->slowkeys_alert);
@@ -644,9 +667,11 @@
                 _("You just pressed two keys at once, or pressed the Shift key 5 times in a row.  "
                   "This turns off the Sticky Keys feature, which affects the way your keyboard works.");
 
+#ifndef HAVE_APPINDICATOR
         if (manager->priv->status_icon == NULL || ! gtk_status_icon_is_embedded (manager->priv->status_icon)) {
                 return FALSE;
         }
+#endif
 
         if (manager->priv->slowkeys_alert != NULL) {
                 gtk_widget_destroy (manager->priv->slowkeys_alert);
@@ -1000,10 +1025,16 @@
                 p->device_manager = NULL;
         }
 
+#ifdef HAVE_APPINDICATOR
+	if (p->app_indicator)
+		app_indicator_set_status (p->app_indicator,
+					  APP_INDICATOR_STATUS_PASSIVE);
+#else
         if (p->status_icon) {
                 gtk_status_icon_set_visible (p->status_icon, FALSE);
                 p->status_icon = NULL;
         }
+#endif
 
         if (p->settings != NULL) {
                 g_signal_handlers_disconnect_by_func (p->settings, keyboard_callback, manager);
@@ -1068,8 +1099,13 @@
 }
 
 static void
+#ifdef HAVE_APPINDICATOR
+on_status_icon_activate (GtkMenuItem            *item,
+                         GsdA11yKeyboardManager *manager)
+#else
 on_status_icon_activate (GtkStatusIcon          *status_icon,
                          GsdA11yKeyboardManager *manager)
+#endif
 {
         if (manager->priv->preferences_dialog == NULL) {
                 manager->priv->preferences_dialog = gsd_a11y_preferences_dialog_new ();
@@ -1102,6 +1138,26 @@
 {
         gnome_settings_profile_start (NULL);
 
+#ifdef HAVE_APPINDICATOR
+	if (!manager->priv->app_indicator) {
+		GtkWidget *menu = gtk_menu_new ();
+		GtkWidget *item = gtk_menu_item_new_with_label (_("Universal Access Preferences"));
+
+		g_signal_connect (item,
+				  "activate",
+				  G_CALLBACK (on_status_icon_activate),
+				  manager);
+
+		gtk_menu_shell_append (GTK_MENU_SHELL (menu), item);
+		gtk_widget_show_all (menu);
+
+		manager->priv->app_indicator = app_indicator_new ("a11y-keyboard",
+								  "preferences-desktop-accessibility",
+								  APP_INDICATOR_CATEGORY_OTHER);
+		app_indicator_set_menu (manager->priv->app_indicator,
+					GTK_MENU (menu));
+	}
+#else
         if (!manager->priv->status_icon) {
 
                 manager->priv->status_icon = gtk_status_icon_new_from_icon_name ("preferences-desktop-accessibility");
@@ -1115,6 +1171,7 @@
                                   G_CALLBACK (on_status_icon_popup_menu),
                                   manager);
         }
+#endif
 
         gnome_settings_profile_end (NULL);
 }
Index: gnome-settings-daemon-3.6.3/plugins/a11y-keyboard/Makefile.am
===================================================================
--- gnome-settings-daemon-3.6.3.orig/plugins/a11y-keyboard/Makefile.am	2012-11-14 13:19:27.000000000 +0100
+++ gnome-settings-daemon-3.6.3/plugins/a11y-keyboard/Makefile.am	2012-11-19 19:54:26.275630058 +0100
@@ -25,6 +25,7 @@
 	-DGTKBUILDERDIR=\""$(pkgdatadir)"\"			\
 	$(PLUGIN_CFLAGS)					\
 	$(LIBNOTIFY_CFLAGS)					\
+	$(APPINDICATOR_CFLAGS)				\
 	$(SETTINGS_PLUGIN_CFLAGS)				\
 	$(AM_CFLAGS)
 
@@ -32,6 +33,7 @@
 	$(top_builddir)/gnome-settings-daemon/libgsd.la		\
 	$(top_builddir)/plugins/common/libcommon.la		\
 	$(LIBNOTIFY_LIBS)					\
+	$(APPINDICATOR_LIBS)				\
 	$(SETTINGS_PLUGIN_LIBS)
 
 test_a11y_preferences_dialog_SOURCES =		\
@@ -79,6 +79,7 @@
 liba11y_keyboard_la_CFLAGS = \
 	$(SETTINGS_PLUGIN_CFLAGS)	\
 	$(LIBNOTIFY_CFLAGS)		\
+	$(APPINDICATOR_CFLAGS)		\
 	$(AM_CFLAGS)
 
 liba11y_keyboard_la_LDFLAGS = 		\
@@ -89,6 +90,7 @@
 	$(SETTINGS_PLUGIN_LIBS)		\
 	$(XF86MISC_LIBS)		\
 	$(LIBNOTIFY_LIBS)		\
+	$(APPINDICATOR_LIBS)		\
 	$(NULL)
 
 plugin_in_files = 		\
