Description: Show in Unity too
Author: Michael Terry <michael.terry@canonical.com>
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=654919
Bug-Ubuntu: https://launchpad.net/bugs/803519

Index: gnome-settings-daemon-3.6.1/plugins/automount/gnome-fallback-mount-helper.desktop.in.in
===================================================================
--- gnome-settings-daemon-3.6.1.orig/plugins/automount/gnome-fallback-mount-helper.desktop.in.in	2012-05-24 11:50:44.000000000 +0200
+++ gnome-settings-daemon-3.6.1/plugins/automount/gnome-fallback-mount-helper.desktop.in.in	2012-10-26 10:19:49.627004182 +0200
@@ -9,4 +9,4 @@
 NoDisplay=true
 OnlyShowIn=GNOME;Unity;
 X-GNOME-Autostart-Notify=true
-AutostartCondition=GNOME3 if-session gnome-fallback
+AutostartCondition=GNOME3 unless-session gnome
