Index: gnome-settings-daemon-3.6.1/plugins/power/gsd-power-manager.c
===================================================================
--- gnome-settings-daemon-3.6.1.orig/plugins/power/gsd-power-manager.c	2012-10-26 10:19:50.127004207 +0200
+++ gnome-settings-daemon-3.6.1/plugins/power/gsd-power-manager.c	2012-10-26 10:19:50.151004207 +0200
@@ -1167,7 +1167,7 @@
 
         /* add to database */
         array = up_client_get_devices (manager->priv->up_client);
-        for (i=0;i<array->len;i++) {
+        for (i = 0; array != NULL && i < array->len; i++) {
                 device = g_ptr_array_index (array, i);
                 engine_device_add (manager, device);
                 engine_check_recall (manager, device);
